clang/lib/AST/ExprConstant.cpp

```diff
@@ -5346,7 +5346,6 @@ static EvalStmtResult EvaluateStmt(StmtResult &Result, EvalInfo &Info,
     const Expr *RetExpr = cast<ReturnStmt>(S)->getRetValue();
     FullExpressionRAII Scope(Info);
     if (RetExpr && RetExpr->isValueDependent()) {
-      EvaluateDependentExpr(RetExpr, Info);
```

Comment:
This is a pure function that's supposed to be called only for its return value, which is being discarded here.

That function has an assert internally that is firing when we transform a declrefexpr referencing a lambda.
This happens on a transform with an incomplete template argument list. Empty even, all template arguments being null.

I can double check for more details, but a transform pass without any actual transformation shouldn't produce errors.

Reply:
Looking at this again, leaving this line would break this test case:

```C++
template<int N> void g(const char (*)[([]{ return N; })()]) {}
template<int N> void g(const char (*)[([]{ return N; })()]) {}
void use_g() { g<6>(&"hello"); }
```

On the partial ordering case, N stays dependent, so we hit here.
This function assumes that, in certain evaluation modes, this expression has to contain an error.
But I don't see how that assumption holds true from the surrounding code.

CC @hokein since he implemented this originally, and was the last one to touch this code.

Reply:
I can't answer that question as I am not even sure what it's supposed to do here.
I can't prove it doesn't do anything at all without spending some time into it, which hopefully we could avoid if @hokein would shed some light :)

Reply:
If I recall correctly, historically (before the introduction of RecoveryExpr), we assumed that a non-template function never contained any value-dependent expressions. The constant expression evaluator relied on this assumption. However, with the introduction of RecoveryExpr (which is always value-dependent), we started encountering value-dependent expressions outside of dependent contexts.

We enhanced the constant evaluator to handle these value-dependent cases, but only specifically for RecoveryExpr. This code path should now only be triggered by RecoveryExpr, which is why we expect the containErrors bit in EvaluateDependentExpr.

The removal of EvaluateDependentExpr seems a bit suspicious to me, as we still call EvaluateDependentExpr in several other places.

With this change, are we starting to see value-dependent expressions that are not RecoveryExpr in the constant expression evaluator?

Reply:
I see. The problem is that in this case we can have incomplete substitutions. For most of the template instantiator, this is not a problem since rebuilding into any dependent stuff just suppresses further analysis, and we just discard the result of the substitution since it was incomplete anyway.

This breaks down here because of the lambda, since a call to an incompletely substituted lambda can come off as non-dependent, even though you would find dependent stuff if you tried to evaluate it.

If the constant expression evaluator would treat finding any dependent stuff as an error, this would come off ok. But since you pointed out there would need to be more places to fix, I think what we could do instead is just bail out immediately when  we know we have an incomplete substitution.